File: wolt_cli/services/output.py
import datetime
import json
import uuid
from enum import Enum
from pathlib import Path
from typing import Any

import typer
from rich.console import Console
from rich.table import Table


class OutputFormat(str, Enum):
    TABLE = "table"
    JSON = "json"
    YAML = "yaml"


def build_envelope(
    *,
    profile: str,
    locale: str,
    data: Any,
    warnings: list[str] | None = None,
    error: dict[str, Any] | None = None,
) -> dict[str, Any]:
    envelope: dict[str, Any] = {
        "meta": {
            "request_id": f"req_{uuid.uuid4().hex}",
            "generated_at": datetime.datetime.now(datetime.UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z"),
            "profile": profile,
            "locale": locale,
        },
        "data": data,
        "warnings": warnings or [],
    }
    if error is not None:
        envelope["error"] = error
    return envelope


def render_payload(payload: dict[str, Any], output_format: OutputFormat) -> str:
    if output_format == OutputFormat.JSON:
        return json.dumps(payload, ensure_ascii=False, indent=2)
    if output_format == OutputFormat.YAML:
        try:
            import yaml
        except ModuleNotFoundError as exc:
            raise RuntimeError("yaml output requires PyYAML to be installed") from exc
        return yaml.safe_dump(payload, sort_keys=False, allow_unicode=True)
    raise ValueError("render_payload() only supports json or yaml output")


def emit_machine_payload(payload: dict[str, Any], output_format: OutputFormat, output_path: Path | None = None) -> None:
    rendered = render_payload(payload, output_format)
    if output_path:
        output_path.write_text(rendered)
    typer.echo(rendered)


def emit_table(table: Table, *, no_color: bool = False, output_path: Path | None = None) -> None:
    console = Console(no_color=no_color, record=output_path is not None)
    console.print(table)
    if output_path:
        output_path.write_text(console.export_text(styles=not no_color))

File: internal/service/output/output.go
package output

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

// Format represents command output encoding.
type Format string

const (
	FormatTable Format = "table"
	FormatJSON  Format = "json"
	FormatYAML  Format = "yaml"
)

// ParseFormat validates format values.
func ParseFormat(v string) (Format, error) {
	switch Format(strings.ToLower(strings.TrimSpace(v))) {
	case "", FormatTable:
		return FormatTable, nil
	case FormatJSON:
		return FormatJSON, nil
	case FormatYAML:
		return FormatYAML, nil
	default:
		return "", fmt.Errorf("unsupported format %q", v)
	}
}

func newRequestID() string {
	buf := make([]byte, 8)
	if _, err := rand.Read(buf); err != nil {
		return "req_fallback"
	}
	return "req_" + hex.EncodeToString(buf)
}

// Envelope is the machine-output payload.
type Envelope struct {
	Meta     map[string]any `json:"meta" yaml:"meta"`
	Data     any            `json:"data" yaml:"data"`
	Warnings []string       `json:"warnings" yaml:"warnings"`
	Error    map[string]any `json:"error,omitempty" yaml:"error,omitempty"`
}

// BuildEnvelope constructs a response envelope.
func BuildEnvelope(profile, locale string, data any, warnings []string, errPayload map[string]any) Envelope {
	env := Envelope{
		Meta: map[string]any{
			"request_id":   newRequestID(),
			"generated_at": time.Now().UTC().Truncate(time.Second).Format(time.RFC3339),
			"profile":      profile,
			"locale":       locale,
		},
		Data:     data,
		Warnings: warnings,
		Error:    errPayload,
	}
	if env.Warnings == nil {
		env.Warnings = []string{}
	}
	return env
}

// RenderPayload renders payload in json/yaml format.
func RenderPayload(payload Envelope, format Format) (string, error) {
	switch format {
	case FormatJSON:
		bytes, err := json.MarshalIndent(payload, "", "  ")
		if err != nil {
			return "", fmt.Errorf("marshal json: %w", err)
		}
		return string(bytes), nil
	case FormatYAML:
		bytes, err := yaml.Marshal(payload)
		if err != nil {
			return "", fmt.Errorf("marshal yaml: %w", err)
		}
		return string(bytes), nil
	default:
		return "", fmt.Errorf("render payload only supports json/yaml")
	}
}

// WriteOutput writes output to the provided writer and optional file.
func WriteOutput(w io.Writer, text string, outputPath string) error {
	if outputPath != "" {
		if err := os.WriteFile(outputPath, []byte(text), 0o644); err != nil {
			return fmt.Errorf("write output file: %w", err)
		}
	}
	if _, err := fmt.Fprintln(w, text); err != nil {
		return fmt.Errorf("write output: %w", err)
	}
	return nil
}

// RenderTable renders plain text tables.
func RenderTable(title string, headers []string, rows [][]string) string {
	var b strings.Builder
	if title != "" {
		b.WriteString(title)
		b.WriteByte('\n')
	}
	if len(headers) > 0 {
		b.WriteString(strings.Join(headers, "\t"))
		b.WriteByte('\n')
	}
	for _, row := range rows {
		b.WriteString(strings.Join(row, "\t"))
		b.WriteByte('\n')
	}
	return strings.TrimRight(b.String(), "\n")
}
